### PR TITLE
Field length validation for qualification type in vendor API

### PR DESCRIPTION
**Pull request by stevehook**
## Context

We have a documented limit of 256 characters on the `ApplicationQualification#qualification_type` field in the Vendor API. However we permit values longer than this getting into the database. This has caused problems for certain vendors because a few candidates have entered some _very_ long qualification types.

## Changes proposed in this pull request

- Data migration to truncate any existing qualification types.
- Form and model validation to prevent future entry of long qualification types.

## Guidance to review

Per commit

## Link to Trello card

[](https://trello.com/c/HsGFLF9O/1231-restrict-field-length-of-qualificationtype-fields-in-vendor-api)

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/services/data_migrations/truncate_long_qualification_type_values.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module DataMigrations
+  class TruncateLongQualificationTypeValues
+    TIMESTAMP = 20230320223629
+    MANUAL_RUN = false
+
+    def change
+      long_qualification_type_qualifications.find_each do |qualification|
```
Comment (carlosmartinez) [inline app/services/data_migrations/truncate_long_qualification_type_values.rb:7]: i think we need to do the same for `non_uk_qualification_type`

File: config/locales/candidate_interface/gcse.yml [hunk 2]
```diff
@@ -121,6 +121,7 @@ en:
           attributes:
             qualification_type:
               blank: Select the type of qualification
+              too_long: Type of qualification must be %{count} characters or fewer
```
Comment (frankieroberto) [inline config/locales/candidate_interface/gcse.yml:124]:
```suggestion
              too_long: Type of qualification must be %{count} characters or less
```

**Review by frankieroberto — COMMENTED** [review]
Can we check how many qualifications this data migration will affect? (And maybe pull out some examples?)

I wonder if we could consider keeping the full text in the database, but just truncating the text when presented in the API?

**Comment by stevehook** [comment]
> Can we check how many qualifications this data migration will affect? (And maybe pull out some examples?)

@frankieroberto There are 11 offending records in the production DB (all `non_uk_qualification_type`) - https://www.apply-for-teacher-training.service.gov.uk/support/blazer/queries/490-qualifications-with-a-longer-than-valid-non-uk-qualification-type

**Comment by stevehook** [comment]
> I wonder if we could consider keeping the full text in the database, but just truncating the text when presented in the API?

I think this would solve the main problem. What do you think @carlosmartinez ?

**Comment by carlosmartinez** [comment]
> > I wonder if we could consider keeping the full text in the database, but just truncating the text when presented in the API?
> 
> I think this would solve the main problem. What do you think @carlosmartinez ?

yes, reasonable

**Comment by carlosmartinez** [comment]
> @stevehook @carlosmartinez thanks. Ideally we'd truncate them in the API with a `…` character (so it's super obvious that there's something missing)? Then providers could look at the full thing in Manage if needed?
> 
> (If this is adding too much code though then it's probably acceptable to just truncate them in the db, given there's only 11?)

I don't think we need to spend too much more time on this. As you say, there are only 11. And any value greater than 256 chars for this field is _Bad Data_, by definition.

**Review by carlosmartinez — APPROVED** [review]

**Pull request merged**